plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/SourceAttacherJob.java

```diff
@@ -54,37 +56,36 @@ protected IStatus run(IProgressMonitor monitor) {
-        IClasspathEntry[] newClasspathEntries = new IClasspathEntry[classpathEntries.length];
-        System.arraycopy(classpathEntries, 0, newClasspathEntries, 0, classpathEntries.length);
-        for (int i = 0; i < newClasspathEntries.length; i++) {
-          IClasspathEntry entry = newClasspathEntries[i];
-          if (entry.getPath().equals(libraryPath)) {
-            IPath sourceArtifactPath = sourceArtifactPathProvider.call();
-            newClasspathEntries[i] = JavaCore.newLibraryEntry(entry.getPath(),
-                                                              sourceArtifactPath,
-                                                              null /* sourceAttachmentRootPath */,
-                                                              entry.getAccessRules(),
-                                                              entry.getExtraAttributes(),
-                                                              entry.isExported());
-          }
-        }
-        LibraryClasspathContainer newContainer =
-            libraryClasspathContainer.copyWithNewEntries(newClasspathEntries);
-        JavaCore.setClasspathContainer(containerPath, new IJavaProject[]{ javaProject },
-                                       new IClasspathContainer[]{ newContainer }, monitor);
-        serializer.saveContainer(javaProject, newContainer);
-      } else {
         logger.log(Level.FINE, Messages.getString("ContainerClassUnexpected",
-                                                  container.getClass().getName(),
-                                                  LibraryClasspathContainer.class.getName()));
+            container.getClass().getName(), LibraryClasspathContainer.class.getName()));
+        return Status.OK_STATUS;
+      };
+
+      LibraryClasspathContainer libraryClasspathContainer = (LibraryClasspathContainer) container;
+      IPath sourceArtifactPath = sourceArtifactPathProvider.call();
+      List<IClasspathEntry> newClasspathEntries = new ArrayList<>();
+
+      for (IClasspathEntry entry : libraryClasspathContainer.getClasspathEntries()) {
+        if (!entry.getPath().equals(libraryPath)) {
+          newClasspathEntries.add(entry);
+        } else {
+          newClasspathEntries.add(JavaCore.newLibraryEntry(
+              entry.getPath(), sourceArtifactPath, null /* sourceAttachmentRootPath */,
+              entry.getAccessRules(), entry.getExtraAttributes(), entry.isExported()));
+        }
       }
+
+      LibraryClasspathContainer newContainer =
+          libraryClasspathContainer.copyWithNewEntries(newClasspathEntries);
+      JavaCore.setClasspathContainer(containerPath, new IJavaProject[]{ javaProject },
+                                     new IClasspathContainer[]{ newContainer }, monitor);
+      serializer.saveContainer(javaProject, newContainer);
+      return Status.OK_STATUS;
+
     } catch (Exception ex) {
```

Comment:
`sourceArtifactPathProvider` is a `Callable` and this `call()` declares Exception to be thrown. We can either replace it with a `com.google.common.base.Supplier, swallow any exception that happens inside and return a `null` or create an interface similar to `Callable` and declare `CoreException` to be thrown.